### PR TITLE
Revert "zfsUnstable: 2.1.0 -> unstable-2021-08-30"

### PR DESCRIPTION
**Pull request by zhaofengli**
From the documentation of `boot.zfs.enableUnstable`:

> The version will have already *passed an extensive test suite*, but it is more likely to hit an undiscovered bug compared to running a released version of ZFS on Linux.

(emphasis mine)

However, this specific version [*fails*](https://github.com/openzfs/zfs/runs/3467198842) the test suite. I also don't think we have [ever bumped zfsUnstable to a non-release and non-rc version](https://github.com/zhaofengli/nixpkgs/commits/0046ed37e1ed4010c68f9d1562853b63318f89a0/pkgs/os-specific/linux/zfs/default.nix). Therefore we should revert this before it hits a channel.

### CODE REVIEW

**Comment by adisbladis** [comment]
> I also don't think we have ever bumped zfsUnstable to a non-release and non-rc version.

Sure we have, you just need to go a bit further back in history:
https://github.com/zhaofengli/nixpkgs/commits/0046ed37e1ed4010c68f9d1562853b63318f89a0?after=0046ed37e1ed4010c68f9d1562853b63318f89a0+69&branch=0046ed37e1ed4010c68f9d1562853b63318f89a0&path%5B%5D=pkgs&path%5B%5D=os-specific&path%5B%5D=linux&path%5B%5D=zfs&path%5B%5D=default.nix

> However, this specific version fails the test suite.

At the time of creating the PR that commit is what the upstream HEAD was at. 
Given that the only diff between https://github.com/openzfs/zfs/commit/b1a1c64313c1cb2450b5fb6c3befe5c9e72a0b3e where the tests pass and https://github.com/openzfs/zfs/commit/3b89d9518df2c7fd747e349873a3d4d498beb20e where they don't is only applicable on FreeBSD I chalked that up to a flaky test suite.

> Therefore we should revert this before it hits a channel.

The current state of the channel is one where zfs is broken with `linuxPackages_latest`. Hardly an ideal scenario.

**Comment by zhaofengli** [comment]
> [...] I chalked that up to a flaky test suite.

> The current state of the channel is one where zfs is broken with linuxPackages_latest. Hardly an ideal scenario.

If you just want to fix the build for 5.14, the prudent thing to do is to cherry-pick specific commits onto the stable branch [which the ZFS maintainers have done](https://github.com/openzfs/zfs/pull/12506) and are in the process of cutting a release. They have actually ran the tests locally and I trust their judgement.

Waiting a few days doesn't hurt anyone, especially since the channel hasn't advanced yet to include 5.14. It's just been 2 days since the release and I don't understand the rush. Those who somehow *need* 5.14 can cherry-pick themselves. Apparently those people don't need Hydra anyways since [the PR to advance linux_testing](https://github.com/NixOS/nixpkgs/pull/130572) has been ignored for months.

**Comment by TredwellGit** [comment]
>The current state of the channel is one where zfs is broken with linuxPackages_latest.

`boot.kernelPackages = config.boot.zfs.package.latestCompatibleLinuxPackages;` is supposed to be set for ZFS not `linuxPackages_latest`. (This needs documentation.)

**Review by TredwellGit — APPROVED** [review]

**Comment by lovesegfault** [comment]
> Waiting a few days doesn't hurt anyone, especially since the channel hasn't advanced yet to include 5.14.

It did hurt me, and unstable-small has picked it up already.

I do agree that our policy should be to cherry-pick on top of the stable release though.

**Comment by TredwellGit** [comment]
>It did hurt me

Please elaborate.

**Comment by TredwellGit** [comment]
Maybe there should be an additional ZFS extra unstable package where Nixpkgs maintainers can cherry-pick, so that zfsUnstable is only what the OpenZFS consider safe to use.

**Review by Ma27 — APPROVED** [review]

**Comment by TredwellGit** [comment]
Is this going to be merged? https://github.com/NixOS/nixpkgs/pull/136299 did not even bump latestCompatibleLinuxPackages.

**Comment by hmenke** [comment]
This thread awfully sounds like

> I deliberately installed something marked as unstable which turned out to be unstable and now I am mad.

If you don't want an unstable version, don't use `zfsUnstable`.

As adisbladis has already mentioned, the test failures are most likely spurious since the patch where they don't pass only applies to FreeBSD.

The reason why `latestCompatibleLinuxPackages` was not bumped is that even the unstable version is *not* officially compatible with Linux 5.14. Please refer to the entry `Linux-Maximum` in https://github.com/openzfs/zfs/blob/master/META. This is the *only* authoritative source on Linux compatibility.

BTW, when the kernel is not compatible, ZFS is only marked as broken, so it will still be possible to build and run it by setting `NIXPKGS_ALLOW_BROKEN=1` or `config.allowBroken = true`, see https://nixos.org/manual/nixpkgs/stable/#sec-allow-broken

I'm therefore voting to close this PR.

**Comment by Ma27** [comment]
I gotta say, I'm not a big fan of polemic arguments such as `I deliberately installed something marked as unstable which turned out to be unstable and now I am mad.`

Please do not forget that we're **not** talking about a user-space application where being unstable can become slightly tedious, but about a file-system. Which is also why I'd strictly advice against setting `config.allowBroken` here.

> If you don't want an unstable version, don't use zfsUnstable.

This isn't necessarily true, on 21.05 we have `zfsUnstable` explicitly set to `2.1.0` because it's needed to have a ZFS working with 5.13 (IIRC), but 2.1.0 has a few small breaking changes which is why zfsStable is still 2.0.x. I agree that the situation isn't ideal of course, but this isn't a universally true.

Also, for a file-system that's part of the official package-set of a Linux distro, I'd expect a certain degree of stability, even though there's a higher risk of problems arising (which is why we rely on upstream's test suite).

**Comment by hmenke** [comment]
My preferred solution would be to simply remove `zfsUnstable` altogether. As far as I can seee it mostly causes confusion (https://github.com/NixOS/nixpkgs/issues/109001, https://github.com/NixOS/nixpkgs/issues/109031) or people are surprised that it's actually unstable.

It's not too hard to pop in an overlay for the unstable version and those users who actually want to try out the unstable version should be more than capable to do so
```nix
{
  nixpkgs.overlays = [ (final: prev: let
    version = "unstable-2021-08-30";
    src = final.fetchFromGitHub {
        owner = "openzfs";
        repo = "zfs" ;
        rev = "3b89d9518df2c7fd747e349873a3d4d498beb20e";
        sha256 = "sha256-YdY4SStXZGBBdAHdM3R/unco7ztxI3s0/buPSNSeh5o=";
      };
  in {
    linuxPackages_latest = prev.linuxPackages_latest.extend (lpFinal: lpPrev: {
      zfs = lpPrev.zfs.overrideAttrs (_: {
        name = "zfs-kernel-${version}";
        inherit src;
        meta.broken = false;
      });
    });
    zfs = prev.zfs.overrideAttrs (_: {
      name = "zfs-user-${version}";
      inherit src;
    });
  }) ];
}
```
Maybe this should be offered as the `boot.zfs.enableUnstable` option in NixOS.

**Comment by L-as** [comment]
AFAICT, `Unstable` in the package name means that the version can refer to the latest revision at the time of updating. I don't see why ZFS should be special in this case. To me it sounds like you want a newer version of ZFS that isn't the newest revision available.

We could add a `zfsNewest` attribute, how is that? (You could also call it `zfsNewer`, since it technically isn't the *newest* revision).

**Comment by Ma27** [comment]
> AFAICT, Unstable in the package name means that the version can refer to the latest revision at the time of updating. I don't see why ZFS should be special in this case. To me it sounds like you want a newer version of ZFS that isn't the newest revision available.

Well, we also do `nixUnstable` updates only if they're perfectly fine. Also, as said in the OP, there's an explicit definition of when zfsUnstable is supposed to be bumped!

However, it doesn't seem to me as if there's a majority supporting this change, so I'm a bit hesitant to merge currently.

**Comment by L-as** [comment]
> Well, we also do `nixUnstable` updates only if they're perfectly fine. Also, as said in the OP, there's an explicit definition of when zfsUnstable is supposed to be bumped!

I completely agree that we shouldn't bump it to a version that clearly doesn't pass the test suite. This is an oversight, which somebody should have noticed before the previous PR for `zfsUnstable` was merged. However, I don't think that that means that an unstable package should use stable revisions. We should bump `zfsUnstable` to e.g. https://github.com/openzfs/zfs/commit/6954c22f35cc8182f76bd18ab259a7d3eaf40b65, for which the tests pass.

However, I still think adding a new `zfsNewer` attribute or similar might be a good idea, but I don't use ZFS myself personally.

**Comment by hmenke** [comment]
Can anyone actually provide a backtrace or similar where the current `zfsUnstable` allegedly “fails”? The longer this goes, the more made up this entire issue feels.

**Comment by L-as** [comment]
If you check the commit on GitHub you can see the CI failed.

**Comment by hmenke** [comment]
The ZTS is basically meaningless for us, because the OpenZFS team just tagged the 2.1.1 release which also fails: https://github.com/openzfs/zfs/runs/3624362268

The NixOS tests on the other hand pass. That should be good enough.

**Comment by L-as** [comment]
> The ZTS is basically meaningless for us, because the OpenZFS team just tagged the 2.1.1 release which also fails: https://github.com/openzfs/zfs/runs/3624362268
> 
> The NixOS tests on the other hand pass. That should be good enough.

I don't think the NixOS tests are something you should rely on for testing FS correctness.
As for the newest stable release, it seems that the test failing is the exact same as the test failing here. Considering it's a release, it likely works, but I don't think it's a good policy to use commits that don't pass CI, when we know almost nothing about the ZFS internals. I don't see why you think the ZTS is basically meaningless for us, so if every single test fails, does that not mean we shouldn't use it? It is much better to be on the safe side and not use the few commits where the CI fails.

**Comment by hmenke** [comment]
Have you read the edit to my previous comment? 2.1.0 also failed the test suite and is currently shipped as `zfsStable`. Sure, if the majority of tests in the ZTS fail, it's time to get suspicious, but otherwise I'm just going to throw your argument back at you: Do you know enough about ZFS internals to make assess its quality? Evidently passing all tests is not even required for major releases.

To reiterate my previous point, if you don't want an unstable ZFS, don't use `zfsUnstable`!

**Comment by hmenke** [comment]
Fixed in https://github.com/NixOS/nixpkgs/pull/138168

**Comment by zhaofengli** [comment]
To repeat [my comment](https://github.com/NixOS/nixpkgs/pull/136340#issuecomment-909780894) above:

> If you just want to fix the build for 5.14, the prudent thing to do is to cherry-pick specific commits onto the stable branch [which the ZFS maintainers have done](https://github.com/openzfs/zfs/pull/12506) and are in the process of cutting a release. They have actually ran the tests locally and I trust their judgement.

They tested it locally, put it in staging for a while, then released it. They had the confidence to do that and I trust their judgement.

I don't get why you have to be so confrontational. We haven't bumped zfsUnstable to a non-rc version since 2018 and I think that's what we should be doing. As you have mentioned, advanced users already have the ability to pop in an overlay and use whatever version they want to use.

**Comment by hmenke** [comment]
You are definitely right that zfsUnstable should have been updated to a commit from the staging branch rather than the master branch, but I do not see this as a compelling reason to completely revert this change. A version bump to the correct branch would have been appreciated better. I also got a bit annoyed by people barging in claiming that there are issues without elaborating or not even using ZFS.

**Pull request closed without merging**
